hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1247,6 +1244,60 @@ private class Emit(
 
         ArrayIteratorTriplet(Code._empty, None, f)
 
+      case x@ArrayLeftJoinDistinct(left, right, l, r, compKey, join) =>
+        // no missing
+        val lelt = coerce[TArray](left.typ).elementType
+        val relt = coerce[TArray](right.typ).elementType
+        val rtyp = coerce[PArray](right.pType)
+
+        val larray = emitArrayIterator(left)
+        val rarray = emitArrayIterator(right).toEmitTriplet(mb, rtyp)
+
+        val rm = mb.newField[Boolean]("join_rarray_m")
+        val rv = mb.newField[Long]("join_rarray_v")
+        val ri = mb.newField[Int]("join_rarray_i")
+        val rlen = mb.newField[Int]("join_rarray_len")
+
+        val lm = mb.newField[Boolean]("join_left_m")
+        val lv = mb.newField("join_left_v")(typeToTypeInfo(lelt))
+
+        val fenv = env.bind(
+          l -> (typeToTypeInfo(lelt), lm.load(), lv.load()),
+          r -> (typeToTypeInfo(lelt), rm.load() || rtyp.isElementMissing(region, rv, ri), rtyp.loadElement(region, rv, ri)))
+
+        val compKeyF = mb.fb.newMethod(typeInfo[Region], typeInfo[Int])
+        val et = new Emit(compKeyF, 1).emit(compKey, fenv)
+        compKeyF.emit(Code(et.setup, et.m.mux(Code._fatal("ArrayLeftJoinDistinct: comp can't be missing"), et.value[Int])))
+        val joinF = mb.fb.newMethod(typeInfo[Region], typeToTypeInfo(relt), typeInfo[Boolean], typeToTypeInfo(join.typ))
+        val jet = new Emit(joinF, 1).emit(Subst(join, Env[IR](r -> In(0, relt))), fenv)
+        joinF.emit(Code(jet.setup, jet.m.mux(Code._fatal("ArrayLeftJoinDistinct: joined can't be missing"), jet.v)))
+
+        val ae = { cont: Emit.F =>
+          val aet = larray.arrayEmitter { (m: Code[Boolean], v: Code[_]) =>
+            Code(
+              lm := m,
+              lv.storeAny(v),
+              Code.whileLoop((ri < rlen) && (coerce[Int](compKeyF.invoke(region)) > 0),
+                ri := ri + 1),
+              cont(
+                lm.load(),
+                ((ri < rlen) && coerce[Int](compKeyF.invoke(region)).ceq(0)).mux(
+                  joinF.invoke(region, rtyp.loadElement(region, rv, ri), rm.load() || rtyp.isElementMissing(region, rv, ri)),
+                  joinF.invoke(region, defaultValue(relt), true))))
+          }
+          val setup = Code(
+            rarray.setup,
+            ri := 0,
+            rm := rarray.m,
+            rm.mux(Code._empty,
+              Code(
+                rv := rarray.value[Long],
+                rlen := rtyp.loadLength(region, rv))))
```

Comment:
Is the default value for `rlen == 0` because that's what `mb.newField[Int]` does?